spec/Idempotency.spec.js

```diff
@@ -0,0 +1,247 @@
+'use strict';
+const Config = require('../lib/Config');
+const Definitions = require('../lib/Options/Definitions');
+const request = require('../lib/request');
+const rest = require('../lib/rest');
+const auth = require('../lib/Auth');
+const uuid = require('uuid');
+
+describe_only_db('mongo')('Idempotency', () => {
+  // Parameters
+  /** Enable TTL expiration simulated by removing entry instead of waiting for MongoDB TTL monitor which
+   runs only every 60s, so it can take up to 119s until entry removal - ain't nobody got time for that */
+  const SIMULATE_TTL = true;
+  // Helpers
+  async function deleteRequestEntry(reqId) {
+    const config = Config.get(Parse.applicationId);
+    const res = await rest.find(
+      config,
+      auth.master(config),
+      '_Idempotency',
+      { reqId: reqId },
+      { limit: 1 }
+    );
+    await rest.del(
+      config,
+      auth.master(config),
+      '_Idempotency',
+      res.results[0].objectId);
+  }
+  async function setup(options) {
+    await reconfigureServer({
+      appId: Parse.applicationId,
+      masterKey: Parse.masterKey,
+      serverURL: Parse.serverURL,
+      idempotencyOptions: options,
+    });
+  }
+  // Setups
+  beforeEach(async () => {
+    if (SIMULATE_TTL) { jasmine.DEFAULT_TIMEOUT_INTERVAL = 200000; }
+    await setup({
+      paths: [
+        "functions/.*",
+        "jobs/.*",
+        "classes/.*",
+        "users",
+        "installations"
+      ],
+      ttl: 30,
+    });
+  });
+  // Tests
+  it('should enforce idempotency for cloud code function', async () => {
+    let counter = 0;
+    Parse.Cloud.define('myFunction', () => {
+      counter++;
+    });
+    const params = {
+      method: 'POST',
+      url: 'http://localhost:8378/1/functions/myFunction',
+      headers: {
+        'X-Parse-Application-Id': Parse.applicationId,
+        'X-Parse-Master-Key': Parse.masterKey,
+        'X-Parse-Request-Id': 'abc-123'
+      }
+    };
+    expect(Config.get(Parse.applicationId).idempotencyOptions.ttl).toBe(30);
+    await request(params);
+    await request(params).then(fail, e => {
+      expect(e.status).toEqual(400);
```

Comment:
Should this read 400 or 159, the error code you defined [here](https://github.com/parse-community/Parse-SDK-JS/blob/22d292549feaa333037ee68ca261bee15fcb5473/src/ParseError.js#L363)

Reply:
No, 400 is the http response code, not the Parse Error code.